MDANSE_GUI/Src/MDANSE_GUI/InputWidgets/TrajectoryFilterWidget.py

```diff
@@ -0,0 +1,1463 @@
+        """Apply the constraint formalised in the lambda
+
+        Parameters
+        ----------
+        value : Any
+            The value of the spinbox.
+        """
+        value = self.to_float(value)
+
+        if (
+            not self.constraint
+            or self.constraint(value)
+            or (not hasattr(self, "initial_value"))
+        ):
+            return
+
+        current = self.to_float(self.value())
+        value_found = False
+        if current < self.initial_value:
+            for x in np.arange(value, self.minimum(), -self.singleStep()):
+                if self.constraint(x):
+                    value_found = True
+                    break
+        else:
+            for x in np.arange(value, self.maximum(), self.singleStep()):
+                if self.constraint(x):
+                    value_found = True
+                    break
+
+        if not value_found:
+            self.setValue(self.initial_value)
+            return
+
+        # Update with constrained value
+        self.setValue(x)
+
+    @staticmethod
+    def to_float(value: Any) -> float:
+        """Convert spinbox value to float if string
+
+        Parameters
+        ----------
+        value : Any
+            The value of the spinbox.
+
+        Returns
+        ----------
+        float
+            The value of the spinbox.
+        """
+        if isinstance(value, str):
+            return float(value)
+        return value
+
+
+class FilterPreferencesGroup(QObject):
+    """Interface for a filter preferences group.
+    Provides a grid layout of settings for a given filter.
+
+    """
+
+    # Signal: emits a dictionary of preferences when settings have been updated
+    _preferences_updated = Signal(dict)
+
+    def __init__(self, render_func: Callable):
+        """
+        Parameters
+        ----------
+        render_func : Callable
+            Filter designer function to call when preferences have been updated.
+            Function re-renders the filter designer graph.
+        """
+        super().__init__()
+
+        # Dictionary mapping preferences to their values
+        self.preferences = {}
+
+        # Grid layout into which the input widgets are placed
+        self.grid = QGridLayout()
+
+        # Dictionary mapping setting name to input widget
+        self.widgets = {}
+
+        # Connection: when the preferences have been updated, re-render the filter designer
+        self._preferences_updated.connect(render_func)
+
+    def store_widget(self, name: str, widget: QWidget) -> None:
+        """Stores a widget in self
+
+        Parameters
+        ----------
+        name : str
+            The name of the instance attribute to be stored, or key string.
+        widget : QWidget
+            The wadget to be stored, or value corresponding to the key.
+
+        Returns
+        -------
+        QWidget:
+            The stored widget
+        """
+        self.widgets.update({name: widget})
```

Comment:
```suggestion
        self.widgets[name] = widget
```
Though not sure why this needs to be a function